solr-search/src/main/java/edu/unc/lib/dl/search/solr/service/ChildrenCountService.java

```diff
@@ -0,0 +1,228 @@
+     * @param principals agent principals
+     * @return count of number of child objects
+     */
+    public long getChildrenCount(BriefObjectMetadata container, AccessGroupSet principals) {
+        SolrQuery solrQuery = createBaseQuery(principals);
+
+        // Add access restrictions to query
+        StringBuilder query = new StringBuilder("*:*");
+        restrictionUtil.add(query, principals);
+        solrQuery.setQuery(query.toString());
+
+        solrQuery.setStart(0);
+        solrQuery.setRows(0);
+        solrQuery.setFacet(true);
+
+        StringBuilder filterQuery = new StringBuilder();
+        addFilter(filterQuery, ANCESTOR_PATH, container.getPath().getSearchValue());
+        solrQuery.addFilterQuery(filterQuery.toString());
+
+        try {
+            QueryResponse queryResponse = executeQuery(solrQuery);
+            return queryResponse.getResults().getNumFound();
+        } catch (SolrServerException e) {
+            throw new SolrRuntimeException(e);
+        }
+    }
+
+    /**
+     * Adds a count of the number of children contained by each object in result
+     * list.
+     *
+     * This count is stored in the containers count map under the "child" key.
+     *
+     * @param containers containers to add child counts to.
+     * @param principals agent principals
+     */
+    public void addChildrenCounts(List<BriefObjectMetadata> containers, AccessGroupSet principals) {
+        addChildrenCounts(containers, principals, CHILD_COUNT, null);
+    }
+
+    /**
+     * Adds a count of the number of children contained by each object in result
+     * list.
+     *
+     * The count will be calculated based on any restrictions from baseQuery if
+     * it is provided. The count will be stored in each containers count map
+     * under the provided countKey.
+     *
+     * @param containers containers to add child counts to.
+     * @param principals agent principals
+     * @param countKey key of the count to store.
+     * @param baseQuery Optional. Starting query that will be used for
+     *            calculating child counts.
+     */
+    public void addChildrenCounts(List<BriefObjectMetadata> containers, AccessGroupSet principals,
+            String countKey, SolrQuery baseQuery) {
+
+        Assert.notNull(containers);
+        if (containers.size() == 0) {
+            return;
+        }
+
+        log.debug("Adding child counts of type {} to result list", countKey);
+
+        SolrQuery solrQuery;
+        if (baseQuery == null) {
+            // Create a base query since we didn't receive one
+            solrQuery = createBaseQuery(principals);
+        } else {
+            // Starting from a base query
+            solrQuery = baseQuery.getCopy();
+            // Make sure we aren't returning any normal results
+            solrQuery.setRows(0);
+            // Remove all facet fields so we are only getting ancestor path
+            if (solrQuery.getFacetFields() != null) {
```

Comment:
can we store the result of `solrQuery.getFacetFields()` in a variable to avoid repeated calls?